core/src/tile/tileManager.cpp

```diff
@@ -376,26 +376,40 @@ void TileManager::enqueueTask(TileSet& _tileSet, const TileID& _tileID,
 
 // create and download raster references (recursively)
 // store these rastertasks in this datasource' task
-void TileManager::loadRasterTasks(std::vector<std::shared_ptr<DataSource>>& rasterSources,
-                                  std::shared_ptr<TileTask>& tileTask, const TileID& tileID) {
+void TileManager::loadRasterTasks(std::vector<std::shared_ptr<DataSource>>& _rasterSources,
+                                  std::shared_ptr<TileTask> _tileTask, const TileID& _tileID) {
 
-    auto& rasterTasks = tileTask->rasterTasks();
-    if (rasterTasks.size() >= rasterSources.size()) { return; }
+    auto& rasterTasks = _tileTask->rasterTasks();
+    if (rasterTasks.size() >= _rasterSources.size()) { return; }
 
 
-    for (size_t index = rasterTasks.size(); index < rasterSources.size(); index++) {
-        auto& rasterSource = rasterSources[index];
-        TileID rasterTileID = tileID;
+    for (size_t index = rasterTasks.size(); index < _rasterSources.size(); index++) {
+        auto& rasterSource = _rasterSources[index];
+        TileID rasterTileID = _tileID;
         if (rasterTileID.z > rasterSource->maxZoom()) {
             rasterTileID = rasterTileID.withMaxSourceZoom(rasterSource->maxZoom());
         }
 
         auto rasterTask = rasterSource->createTask(rasterTileID);
 
-        if (!rasterTask->hasRaster() && m_loadPending < MAX_DOWNLOADS) {
+        if (rasterTask->hasRaster()) {
             rasterTasks.push_back(rasterTask);
 
-            if (rasterSource->loadTileData(std::move(rasterTask), m_rasterCallback, true)) {
+            if (_tileTask->hasTile()) {
+                m_workers.enqueue(std::move(_tileTask));
+            }
+
+        } else if (m_loadPending < MAX_DOWNLOADS) {
+            rasterTasks.push_back(rasterTask);
+            // Load tile if it is not in raw cache.
+
+            if (rasterSource->loadTileData(std::move(rasterTask),
+                                           TileTaskCb{[this, _tileTask](auto rasterTask) {
+                                                   auto task = _tileTask;
+                                                   if (task->hasTile()) {
+                                                       m_workers.enqueue(std::move(task));
```

Comment:
This simplifies the dependent task callback to be just a bool (true).. to
make sure tilemanager update is run whenever a dependent download finishes.
On May 2, 2016 6:06 PM, "Varun Talwar" <varun@mapzen.com> wrote:

> Brainstorming. ... while walking now
>
> Dependent rasters need not be sent to workers at all.
>
> Whenever geom task is done the tile is added to this task.
> Tilemanager on its next update can add rafter textures to this built tile
> in the task if it's corresponding dependent raster tasks are done
> (hasRaster) and then move this tile to entry and delete the main geom task.
>
> If in the next tilemanager update, rasters for this task are not done then
> the task  continues to hold the built tile (task not deleted) and is
> checked in next tilemanager update.
> On May 2, 2016 5:51 PM, "Hannes Janetzek" <notifications@github.com>
> wrote:
>
>> In core/src/tile/tileManager.cpp
>> <https://github.com/tangrams/tangram-es/pull/715#discussion_r61811182>:
>>
>> >              rasterTasks.push_back(rasterTask);
>> >
>> > -            if (rasterSource->loadTileData(std::move(rasterTask), m_rasterCallback, true)) {
>> > +            if (_tileTask->hasTile()) {
>> > +                m_workers.enqueue(std::move(_tileTask));
>> > +            }
>> > +
>> > +        } else if (m_loadPending < MAX_DOWNLOADS) {
>> > +            rasterTasks.push_back(rasterTask);
>> > +            // Load tile if it is not in raw cache.
>> > +
>> > +            if (rasterSource->loadTileData(std::move(rasterTask),
>> > +                                           TileTaskCb{[this, _tileTask](auto rasterTask) {
>> > +                                                   auto task = _tileTask;
>> > +                                                   if (task->hasTile()) {
>> > +                                                       m_workers.enqueue(std::move(task));
>>
>> umm, I think there is some room to simplify this :)
>>
>> raster tasks can run on the worker completely independent of the geom
>> task. when any task finishes it requests a redraw. then the geom-tile can
>> pick their raster textures in the next TileManager update.. tomorrow more
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/tangrams/tangram-es/pull/715/files/76c5e007878847342214ff429988fc62de483529..cf0c8f4884517a9bafdcaedb3b0baac4a30a031e#r61811182>
>>
>